pkg/crc/machine/delete.go

```diff
@@ -20,5 +25,61 @@ func (client *client) Delete() error {
 	if err := libMachineAPIClient.Remove(client.name); err != nil {
 		return errors.Wrap(err, "Cannot remove machine")
 	}
+
+	if err := cleanKubeconfig(getGlobalKubeConfigPath(), getGlobalKubeConfigPath()); err != nil {
+		logging.Warn(err)
+	}
 	return nil
 }
+
+func cleanKubeconfig(input, output string) error {
+	cfg, err := clientcmd.LoadFromFile(input)
+	if err != nil {
+		return err
+	}
+
+	var clusterNames []string
+	for name, cluster := range cfg.Clusters {
+		if cluster.Server == fmt.Sprintf("https://api%s:6443", constants.ClusterDomain) {
+			clusterNames = append(clusterNames, name)
+		}
+	}
+	var contextNames []string
+	authNames := make(map[string]struct{})
+	for name, context := range cfg.Contexts {
+		if contains(clusterNames, context.Cluster) {
+			contextNames = append(contextNames, name)
+			authNames[context.AuthInfo] = struct{}{}
+		}
+	}
+	// keep auth if it is shared with other contexts
+	for _, context := range cfg.Contexts {
+		if !contains(clusterNames, context.Cluster) {
+			delete(authNames, context.AuthInfo)
+		}
+	}
```

Comment:
@guillaumerose I didn't get why this block is needed, in above block we are already checking `if contains(clusterNames, context.Cluster)` and if it exist then only add `authNames` and in this block we are going to use `if !contains(clusterNames, context.Cluster)` and try to delete the respective `auth` from `authNames` . Just not able to compile it in my head :(

Reply:
Take this kubeconfig:

``` 
- context1: user1 and cluster1
- context2: user1 and crc cluster
``` 

Without this extra loop, user1 will be deleted because user1 belongs to crc cluster. But this user is needed for context1. If we remove it, it will break context1. 
I must admit I don't know if this can happen.

Reply:
@guillaumerose This can possible as config option but I don't think you can have same auth associated with 2 different cluster because each auth again associated with token which I don't think same for different cluster.

```
- context1: user1 and cluster1 
- context2: user1 and crc cluster
```

The way this can be achived.
```
$ oc config get-contexts 
CURRENT   NAME                                      CLUSTER                AUTHINFO                       NAMESPACE
          /api-crc-testing:6443/developer           api-crc-testing:6443   developer                      
          crc-admin                                 api-crc-testing:6443   kubeadmin                      default
          crc-developer                             api-crc-testing:6443   developer                      default
          kind-reference                            kind-reference         kind-reference                 
          minishift                                 192-168-42-71:8443     developer/192-168-42-71:8443   myproject
*         project1/api-crc-testing:6443/developer   api-crc-testing:6443   developer                      project1

$ oc config set-context test --cluster=192-168-42-71:8443 --user=kubeadmin 
Context "test" created.

$ oc config get-contexts 
CURRENT   NAME                                      CLUSTER                AUTHINFO                       NAMESPACE
          /api-crc-testing:6443/developer           api-crc-testing:6443   developer                      
          crc-admin                                 api-crc-testing:6443   kubeadmin                      default
          crc-developer                             api-crc-testing:6443   developer                      default
          kind-reference                            kind-reference         kind-reference                 
          minishift                                 192-168-42-71:8443     developer/192-168-42-71:8443   myproject
*         project1/api-crc-testing:6443/developer   api-crc-testing:6443   developer                      project1
          test                                      192-168-42-71:8443     kubeadmin                      

```